File: c#/Structural/Proxy/Proxy/Proxy.cs
using Interface.IDocument;
using Audit.Logger;

namespace Proxy.Proxy
{
    public enum UserRoles
    {
        Admin,
        Dr,
        Patient,
    }
    public class ProxyDocument : IDocument
    {
        private readonly string fileName;
        private Dictionary<string, string> authorizedUser = new Dictionary<string, string>();
        private Document.Document? realDocument;

        public ProxyDocument(string fileName, Document.Document realDocument, Dictionary<string, string> authorizedUser)
        {
            this.fileName = fileName;
            this.authorizedUser = authorizedUser;
            this.realDocument = realDocument;
        }

        public async Task Display(string user)
        {
            var logger = new Logger();

            logger.LogAccess(user, fileName);

            if (!Enum.TryParse<UserRoles>(user, out _))
            {
                throw new Exception($"Invalid user role: {user}");
            }

            if (!authorizedUser.ContainsKey(user))
            {
                throw new Exception($"User {user} does not have permission to access this document");
            }

            if (realDocument == null)
            {
                realDocument = new Document.Document(fileName, null);
            }

            if (realDocument != null)
            {
                await realDocument.LoadFromDisk();
                await realDocument.Display(user);
            }

        }
    }
}

File: typescript/src/structural/proxy/proxy/Proxy.ts
import { Document, RealDocument } from "./Document";
import { logAccess } from "../audit/Logger";

enum USER_ROLES {
    Admin = "admin",
    Dr = "Dr.House",
    Patient = "Patient"
}

export class ProxyDocument implements Document{
    private filename: string
    private authorizedUser : Record<string, string>
    private realDocument: RealDocument | null

    constructor(filename: string, realDocument: null, authorizedUser: Record<string, string>){
        this.filename = filename
        this.authorizedUser = authorizedUser
        this.realDocument = realDocument
    }

    async display(user: string): Promise<void> {
        logAccess(user, this.filename);

        if (!Object.values(USER_ROLES).includes(user as USER_ROLES)) {
            throw new Error(`Invalid user role: ${user}`);
        }

        if (!Object.keys(this.authorizedUser).includes(user)) {
            throw new Error(`${user} does not have permission to access this document.`);
        }

        if (this.realDocument === null) {
            this.realDocument = new RealDocument(this.filename, null);
        }

        if (this.realDocument) {
            await this.realDocument.loadFromDisk(); // Esperar a que el documento se cargue
            this.realDocument.display(user);
        }
    }
}